lxd/migrate.go

```diff
@@ -357,6 +357,65 @@ func snapshotToProtobuf(c container) *Snapshot {
 	}
 }
 
+// Check if CRIU supports pre-dumping and number of
+// pre-dump iterations
+func (s *migrationSourceWs) checkForPreDumpSupport() (bool, int) {
+	// Ask CRIU if this architecture/kernel/criu combination
+	// supports pre-copy (dirty memory tracking)
+	criuMigrationArgs := CriuMigrationArgs{
+		cmd:          lxc.MIGRATE_FEATURE_CHECK,
+		stateDir:     "",
+		function:     "feature-check",
+		stop:         false,
+		actionScript: false,
+		dumpDir:      "",
+		preDumpDir:   "",
+		features:     lxc.FEATURE_MEM_TRACK,
+	}
+	err := s.container.Migrate(&criuMigrationArgs)
+
+	if err != nil {
+		// CRIU says it does not know about dirty memory tracking.
+		// This means the rest of this function is irrelevant.
+		return false, 0
+	}
+
+	// CRIU says it can actually do pre-dump. Let's set it to true
+	// unless the user wants something else.
+	use_pre_dumps := true
+
+	// What does the configuration say about pre-copy
+	tmp := s.container.ExpandedConfig()["migration.incremental.memory"]
+
+	if tmp != "" {
+		use_pre_dumps = shared.IsTrue(tmp)
+	}
+	logger.Debugf("migration.incremental.memory %d", use_pre_dumps)
+
+	var max_iterations int
+
+	// migration.incremental.memory.iterations is the value after which the
+	// container will be definitely migrated, even if the remaining number
+	// of memory pages is below the defined threshold.
+	tmp = s.container.ExpandedConfig()["migration.incremental.memory.iterations"]
+	if tmp != "" {
+		max_iterations, _ = strconv.Atoi(tmp)
```

Comment:
Hm, we should do error checking here. And also validate that `max_iterations` is not `< 0`.

Reply:
I initially had the check for < 0 included until I saw that it will not be < 0 due to ```IsUint32```

Reply:
In the case of an Atoi() error should I just set it to 10?

Reply:
No, I would report an error to the user that this is an invalid value. Actually, the property should already be pre-validated in the container config check but I'd still feel better if the error was checked.

> I initially had the check for < 0 included until I saw that it will not be < 0 due to IsUint32

That validated in the container config checks for the property?